src/Microsoft.AspNet.Mvc.Core/ModelBinding/DictionaryModelBinder.cs

```diff
@@ -14,14 +17,86 @@ namespace Microsoft.AspNet.Mvc.ModelBinding
     public class DictionaryModelBinder<TKey, TValue> : CollectionModelBinder<KeyValuePair<TKey, TValue>>
     {
         /// <inheritdoc />
+        public override async Task<ModelBindingResult> BindModelAsync([NotNull] ModelBindingContext bindingContext)
+        {
+            var result = await base.BindModelAsync(bindingContext);
+            if (result == null || !result.IsModelSet)
+            {
+                // No match for the prefix at all.
+                return result;
+            }
+
+            var model = result.Model as Dictionary<TKey, TValue>;
+            Debug.Assert(model != null);
+            if (model.Count != 0)
+            {
+                // ICollection<KeyValuePair<TKey, TValue>> approach was successful.
+                return result;
+            }
+
+            var enumerableValueProvider = bindingContext.ValueProvider as IEnumerableValueProvider;
```

Comment:
By "filtered value provider" do you mean the special cases around `IBindingSourceValueProvider`? If so, that's a orthogonal concept related to choosing one or more value providers that support a particular `BindingSource`. `IEnumerableValueProvider` is about just `GetKeysFromPrefixAsync()` support, once you've got your value provider(s).

This check fails with any value provider that does not implement `IEnumerableValueProvider`. `ReadableStringCollectionValueProvider` (for form and query values) works but nothing else does. The `CompositeValueProvider` defers to the first `IEnumerableValueProvider` that returns keys but does not supplement inner providers that lack `IEnumerableValueProvider` support.

As it happens `ReadableStringCollectionValueProvider` also implements `IBindingSourceValueProvider`.